src/server/index.js

```diff
@@ -201,28 +227,47 @@ class Server {
       const wrappedStream = Wrap(stream, { lengthDecoder: int32BEDecode, lengthEncoder: int32BEEncode })
       const message = await wrappedStream.readLP()
 
-      try {
-        // Try register
-        const { random128: random, peerID } = JoinInit.decode(message.slice())
+      // Try register
+      const { random128: random, peerID } = JoinInit.decode(message.slice())
+
+      if (random && peerID) {
+        try {
+          this.metrics.joinsTotal.inc()
 
-        if (random && peerID) {
           const id = await PeerId.createFromJSON(peerID)
 
           await this._register(random, id, wrappedStream, stream)
           this.addToNetwork(connection, wrappedStream, id)
-        } else {
-          const { target } = DialRequest.decode(message.slice())
-          const targetB58 = new PeerId(target).toB58String()
-
-          log('dial from %s to %s', connection.localPeer.toB58String(), targetB58)
-          await this._dial(targetB58, wrappedStream, stream)
+          this.metrics.joinsSuccessTotal.inc()
+        } catch (error) {
+          log(error)
+          wrappedStream.writePB({ error: Error.E_GENERIC }, JoinVerify) // if anything fails, respond with generic error
```

Comment:
this should be an outer catch, so any error during this procedure gets caught or everything outside it needs to be unable to throw